Add SEVSNP plugin for Veraison server

jraman567 opened the pull request:
This PR implements the SEVSNP scheme for Veraison.

yogeshbdeshpande commented on scheme/sevsnp/README.md:
```diff
@@ -0,0 +1,58 @@
+# SEV-SNP scheme
+
+This scheme supports the provisioning of reference values and appraisal of evidence. It is suitable for anyone performing verification of simple SEV-SNP evidence.
```
```suggestion
This scheme supports the provisioning of Reference Values and appraisal of Evidence. It is suitable for anyone performing verification of SEV-SNP Evidence.
```

yogeshbdeshpande commented on scheme/sevsnp/README.md:
````diff
@@ -0,0 +1,58 @@
+cocli corim submit --corim-file=scheme/sevsnp/test/refval-prov.cbor --api-server="https://localhost:9443/endorsement-provisioning/v1/submit" --media-type="application/corim-unsigned+cbor; profile=\"https://amd.com/ark\""
+```
+
+### Submitting evidence
````
```suggestion
### Submitting Evidence
```

yogeshbdeshpande commented on scheme/sevsnp/common.go:
```diff
@@ -0,0 +1,199 @@
+}
+
+// transformSVNtoTCB extracts TCB from the supplied SVN. SEV-SNP's TCB_VERSION
+// is a composite version; it's bitfield consisting of SVNs from various firmware components
```
```suggestion
// is a composite version; it's a bitfield consisting of SVNs from various firmware components
```

yogeshbdeshpande commented on scheme/sevsnp/common.go:
```diff
@@ -0,0 +1,199 @@
+		tcbParts   kds.TCBParts
+	)
+
+	// ToDo: following is a circuitous way to obtain the 64-bit TCB integer value
```
If this is a genuine TO DO, preferably create an issue and tag the issue here, please!

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+		return nil, err
+	}
+
+	refValCorim := corim.UnsignedCorim{}
```
just check, if we need to initialise other mandatory part of CoRIM here as well...

jraman567 replied:
I'm setting the Profile and CoMID fields below. Did I miss any other mandatory fields?

Should I use the NewUnsignedCorim() method here instead?

yogeshbdeshpande replied:
Just rename refValCorim to evCoRIM

yogeshbdeshpande reviewed (commented):
in middle of review, just started today

yogeshbdeshpande commented:
As we are adding new `scheme`, please refer to https://github.com/veraison/services/blob/main/README.md

- Insert Sev-snp support in the suitable place for the text pertaining to Provisioning
- Insert Sev-snp support in the suitable place for the text pertaining to Verification

Also remember to check the suitable places in `docs` repo and the Veraison book, to check where this comment is applicable.

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+	result := handler.CreateAttestationResult(SchemeName)
+
+	appraisal := result.Submods[SchemeName]
+
```
Just a question: Is this init to default values or more than this..?
Can you please specify, perhaps a short comment might help as well!

jraman567 replied:
@yogeshbdeshpande Thanks for your feedback, but I wasn't sure how to address it.

This one isn't initializing. The initialization of the Trustworthiness Vector happens below. Could you kindly provide more details on what I should do here? Thanks!

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+		// - POLICY ToDo: Do we need to test individual policy features?
+		// - CURRENT_TCB is informational only. It's best handled by policy
+		// - PLATFORM_INFO ToDO: Do we need to test individual platform features?
+		// - REPORT_DATA is a nonce supplied by user for freshness, not applicable for verification
```
```suggestion
		// - REPORT_DATA is a nonce supplied by user for freshness,  this is verified prior at a suitable point- functon-name() 
```

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+		// - REPORT_DATA is a nonce supplied by user for freshness, not applicable for verification
+		// - REPORT_ID is ephemeral, so we can't use it for verification.
+		// - REPORT_ID_MA is also ephemeral, used for migration
+		// - CHIP_ID is unique to an specific attester, but reference values could be used more generally
```
Just a question:

Does the CHIP_ID maps to a specific instance of platform??? If Yes, shall we not populate the InstanceID Claims..

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+		//     ToDo: It is a good idea to test it anyway, but the Version type only tests for
+		//     equality, and this would trigger spurious failures
+		// - COMMITTED_TCB is informational, used by the host to advance REPORTED_TCB
+		if k == mKeyPolicy ||
```
I have a question here. Should we not expect a definite set of MKeys and check for their presence here or somewhere earlier to report  a fail, if a mandatory MKey is missing from Evidence?
Does this happen intrinsically inside the Evidence Decoding, then it is fine, but just checking...

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+				break claimsLoop
+			}
+			if !compareTcb(*reportedTcb, *em) {
+				// ToDo: Is this a failure condition?
```
Note: Each of these failures should be represented in the corresponding flag inside the TrustVector for Platform ? If not please check and raise a bug on AR4SI if something fundamental missing in the spec..

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler_test.go:
```diff
@@ -0,0 +1,127 @@
+	0x1f, 0x1e, 0x1d, 0x1c, 0x1b, 0x1a, 0x19, 0x18,
+}
+
+func Test_ExtractClaims_ok(t *testing.T) {
```
Just checking, I did not see the Full Evidence JSON Claims set in the test cases..?

is there any reason for this..?

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler_test.go:
```diff
@@ -0,0 +1,127 @@
+	require.NoError(t, err)
+}
+
+func Test_ValidateEvidenceIntegrity_ok(t *testing.T) {
```
Also, I saw three types of Evidence Format in the code, but I see ONLY tests with RATSD Token only,

Can you please add the test cases for TSM Report in JSON and CBOR as well ?

yogeshbdeshpande reviewed (commented):
Left some further comments.

thomas-fossati commented on scheme/sevsnp/common.go:
```diff
@@ -0,0 +1,199 @@
+func parseCertificateChainFromEvidence(tsm *tokens.TSMReport) (*sevsnp.CertificateChain, error) {
+	var certTable abi.CertTable
+
+	if err := certTable.Unmarshal(tsm.AuxBlob); err != nil {
```
Should we first make sure that AuxBlob is not nil?

jraman567 replied:
OK, I added this check. Thanks

thomas-fossati commented on scheme/sevsnp/common.go:
```diff
@@ -0,0 +1,199 @@
+
+		cmwBase64, ok := eat["cmw"].(string)
+		if !ok {
+			return nil, fmt.Errorf("cmw not found in eat")
```
```suggestion
			return nil, errors.New("CMW not found in EAT")
```

thomas-fossati commented on scheme/sevsnp/common.go:
```diff
@@ -0,0 +1,199 @@
+			return nil, err
+		}
+
+		cmwCollection, err := cmw.NewCollection("tag:github.com,2025:veraison/ratsd/cmw")
+		if err != nil {
+			return nil, err
+		}
```
This is the decoding path, so I am not sure why we need to allocate the collection upfront rather than just declaring a variable of type cmw and letting Unmarshal handle allocation.

thomas-fossati commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+		attestation          sevsnp.Attestation
+	)
+
+	options := verify.Options{
```
IIUC, CRL checking is controlled by `CheckRevocations` and is disabled by default.

Would it be worth checking?

jraman567 replied:
Thanks for pointing it out, @thomas-fossati!

We would need a caching server to accomplish this effectively. That's because, AMD's KDS will rate-limit repeated requests for CRL.

thomas-fossati commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+// compareMeasurements checks if two given comid.Measurement variables are the same.
+func compareMeasurements(refM comid.Measurement, evM comid.Measurement) bool {
+	// RawValue comparison
+	if refM.Val.RawValue != nil {
```
Could you please help me understand the matching logic?  I would have expected Evidence to drive the process.  In other words, if evidence reports a raw value, digest, or SVN, we should attempt to match it against the reference values.  However, the logic is inverted here, which is confusing.

jraman567 replied:
Please go one step above the stack to `AppraiseEvidence` method and check `claimsLoop`.

The principle here is evidence must have all the claims in the Reference Value. Evidence can have extra claims, but it must _**at the least**_ have all the claims in the RV.

We ignore the extra claims not present in RV. Does that sound good?

thomas-fossati commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+	return &evComid.Triples.ReferenceValues.Values[0], nil
+}
+
+// compareMeasurements checks if two given comid.Measurement variables are the same.
```
The comment is slightly misleading: the function does not check whether the two comid.Measurement variables are the "same", only whether one of the LHS attributes (picked in a specific order) is the same as the RHS's.

thomas-fossati commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,530 @@
+
+func compareTcb(refM comid.Measurement, evM comid.Measurement) bool {
+	if refM.Val.SVN == nil {
+		log.Errorf("reference doesn't have SVN")
```
```suggestion
		errors.New("reference doesn't have SVN")
```

thomas-fossati replied:
However, just a high-level comment on the API signature: returning an error rather than just a boolean provides richer semantics to the caller.

thomas-fossati reviewed (commented):
Thanks, this is great.

I have left a few comments inline.

jraman567 commented on scheme/sevsnp/common.go:
```diff
@@ -0,0 +1,199 @@
+	return block.Bytes, nil
+}
+
+func parseAttestationToken(token *proto.AttestationToken) (*tokens.TSMReport, error) {
```
Hi @yogeshbdeshpande, are you asking where they are documented?

tsm-report's CDDL is here: https://github.com/veraison/ratsd/blob/main/docs/tsm-report.cddl

1. TSM in CBOR is recorded here: https://www.iana.org/assignments/media-types/application/vnd.veraison.tsm-report+cbor

I couldn't find the IANA doc for the other two. I'll check with @thomas-fossati on how to get them.

Besides we have this godoc: https://pkg.go.dev/github.com/veraison/ratsd/attesters/tsm

jraman567 commented:
> * Also, I feel somewhere in the documentation of the Scheme, we should refer to the documentation on the Evidence Profile and also preferably the Reference Value Profile document, we have used to build this scheme!

OK, I'll add this to the scheme's README.

jraman567 commented on scheme/sevsnp/README.md:
```diff
@@ -0,0 +1,58 @@
+Following is an example of how to interface with this scheme/plugin. The workflow involves using cocli to submit reference values and ratsd to submit the evidence.
+
+Since ratsd is under construction, please use the following instance of evcli to submit evidence.
+https://github.com/jraman567/evcli
```
@thomas-fossati Hey Thomas, I'm changing this to use ratsd instead.

jraman567 commented:
Hi @yogeshbdeshpande and @thomas-fossati , I've addressed most of your comments. The ones I needed more clarification aren't resolved.

Please let me know about the unresolved, so I can address them and close out this PR. Thank you so much! :)

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,557 @@
+	}
+
+	if evM.Val.SVN == nil {
+		log.Errorf("evidence doesn't have SVN")
```
```suggestion
		errors.New("evidence doesn't have SVN")
```

jraman567 commented:
> > Hi @yogeshbdeshpande and @thomas-fossati , I've addressed most of your comments. The ones I needed more clarification aren't resolved.
> > Please let me know about the unresolved, so I can address them and close out this PR. Thank you so much! :)
> 
> I have a couple of comments that should probably be acted upon. They are both very simple.

I'm working on your comments, @thomas-fossati . Will update the PR shortly. Thank you!

yogeshbdeshpande commented on scheme/sevsnp/evidence_handler.go:
```diff
@@ -0,0 +1,557 @@
+
+	appraisal := result.Submods[SchemeName]
+
+	appraisal.TrustVector.InstanceIdentity = ear.NoClaim
```
```suggestion
   // init the TrustVector to default values
	appraisal.TrustVector.InstanceIdentity = ear.NoClaim
```

yogeshbdeshpande commented on scheme/sevsnp/README.md:
````diff
@@ -0,0 +1,68 @@
+cocli corim submit --corim-file=scheme/sevsnp/test/refval-prov.cbor --api-server="https://localhost:9443/endorsement-provisioning/v1/submit" --media-type="application/corim-unsigned+cbor; profile=\"https://amd.com/ark\""
+```
+
+### Submitting evidence
````
```suggestion
### Submitting Evidence
```

yogeshbdeshpande commented:
@thomas-fossati : Please give a final look and provide your approval...

jraman567 commented:
Thank you so much for reviewing, @yogeshbdeshpande and @thomas-fossati !
